raft/src/main/java/org/apache/kafka/snapshot/SnapshotWriter.java

```diff
@@ -78,6 +81,52 @@ public SnapshotWriter(
         );
     }
 
+    /**
+     * FIXME ngoel Improve this comment
+     * Instantiates a SnapshotWriter and stamps the underlying
+     * snapshot with a @MetadataSnapshotHeaderRecord.
+     *
+     * @return {@link Optional}{@link SnapshotWriter}
+     */
+    public static <T> Optional<SnapshotWriter<T>> create(
+            Supplier<Optional<RawSnapshotWriter>> supplier,
+            int maxBatchSize,
+            MemoryPool memoryPool,
+            Time snapshotTime,
+            CompressionType compressionType,
+            RecordSerde<T> serde
+            ){
+        Optional<SnapshotWriter<T>> writer = supplier.get().map(snapshot -> {
+            return new SnapshotWriter<T>(
+                    snapshot,
+                    maxBatchSize,
+                    memoryPool,
+                    snapshotTime,
+                    CompressionType.NONE,
+                    serde
+                    );
+        });
+        if (writer.isPresent()) {
+            //TODO ngoel Append Header
+            // Once the snapshot is intantiated assert that
+            // the snapshot file is 0 bytes and then append a
+            // header
+            SnapshotWriter<T> snapWriter = writer.get();
+            RawSnapshotWriter snapshot = snapWriter.snapshot;
+            if (snapshot.sizeInBytes() != 0){
+                throw new IllegalStateException("Initializing new snapshot with a non-empty file");
+            }
+            // Append the header control record.
+            // FIXME ngoel Where should we define the version?
+            short headerVersion = 0;
+            MetadataSnapshotHeaderRecord headerRecord = new MetadataSnapshotHeaderRecord().setVersion(headerVersion);
+            snapWriter.accumulator.appendSnapshotHeaderMessage(headerRecord);
+            snapWriter.accumulator.forceDrain();
```

Comment:
I would move all of this functionality as a private instance method of `SnapshotWriter`.

Reply:
Done